lib/lcdbwf/R/functional_enrichment.R

```diff
@@ -121,13 +126,19 @@ enrich_test <- function(res, TERM2GENE, TERM2NAME, config, direction, kind='OR',
       direction=direction,
       return_type="rownames"
     )
-
+    # Update the `universe` variable to a vector of background or all genes
+    if (universe) {
+      universe <- rownames(res$res) # uses only detected genes in the current dataset
```

Comment:
For this line, if I understand correctly, "detected" is being defined as "everything in the results object" which ultimately corresponds to, "every annotated gene seen by featureCounts".

Or is the idea that you would manually subset `res$res` before sending it here to only include, say, `res$res[rowSums(res$res)>0,]`?

Maybe we need a way of specifying a list of genes to be passed in here. Or maybe do the `rowSums > 0` within this function? Or perhaps filter out anything with [`isna(padj)`](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA) to get rid of genes that were not even considered by multiple testing correction?

Reply:
My design was the same as what's been described here:

>  "detected" is being defined as "everything in the results object" which ultimately corresponds to, "every annotated gene seen by featureCounts".

Assume there are 8 samples: WT rep1-4, KO rep1-4. Then, there are 3 gene groups here:

a. genes with raw counts > 0 in all samples
b. genes with raw counts > 0 in some samples 
c. genes with raw counts > 0 in no samples

_Raw counts_ mean the number of read counts annotated by featureCounts and my `if` statement was designed to choose genes belonging to 'a' and 'b'. Checking [DESeq2 documentation](https://bioconductor.org/packages/3.7/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA), I assumed that the `baseMean` and `log2FoldChange` columns would return `NA` if a gene is not counted in all samples ('c'). At least within my projects, no `NA` was found in the `baseMean` and `log2FoldChange` columns. 

Regarding this issue, I can bring up two scenarios:

- scenario 1: genes with `NA`s in the `baseMean` and `log2FoldChange` columns were already removed before being saved in the `combined.Rds` file in the `rnaseq.Rmd` script
- scenario 2: coincidently my dataset didn't have `NA`s in those two columns

I thought this case was the scenario 1 but I couldn't find the filtering step when looked into `rnaseq.Rmd`, `helpers.R`, `contrasts.R`, and `functional-enrichment.R`. It's still unclear. Adding a filtering step sounds like a great idea if filtering was not performed in the previous script (e.g. `rnaseq.Rmd`).

Comparing total detected genes vs genes considered by multiple hypothesis testing, I would like to go with total detected genes, because they will be used as _background_ of overrepresentation analysis, conceptually. Here, I considered biological background rather than statistical background. I'd like to hear your thoughts! :)

Reply:
I think my broader question is, should the `universe` argument be a boolean or a list of genes?

Currently it's bool, so it's up to the user to construct the results object appropriately, and that should be more clearly indicated in the help for the function as well as in the config file.

If `universe` should instead be a list of genes, then a helper function would be nice, and that could allow the user to choose what "background" means: any genes with nonzero counts in any sample, or any genes with a valid `padj` (which would include genes with zero counts as well as those with too-low to consider).

I'm leaning towards the list-of-genes approach because it's more explicit. However it would need some more thinking about how to handle it in the config file. Maybe the `limit_background_genes: TRUE` should instead allow different methods of choosing background? And if FALSE then the default universe (all annotated genes) is used.

Reply:
I got your point. It sounds reasonable to me, too! 

How do these 4 categories look in the `config.yaml` file? 

```yaml
  # This sets the `universe` argument of the `enricher` function to choose an option how to limit
  # background genes. Choose one from the following options:
  # - positive_counts_all: genes with positive raw counts in all samples
  # - positive_counts_some: genes with positive raw counts in one or more samples
  # - valid_padj: genes analyzed in multiple hypothesis testing. In DESeq2, genes with very
  #   low counts have high dispersion. These genes are not considered to be significantly different
  #   between groups, and labeled NA in padj.
  # - all_in_DB: disabling to limit background genes. This option will use all available genes in each DB.
  limit_background_genes: "positive_counts_all"
```

I have terrible naming sense. Feel free to suggest better names. 
If they look ok, I'll go related functions based on them.

Reply:
How about the following? I tried to be very clear about what's being excluded. Also, since the background should be "anything that had a chance of being detected", I think the NA/padj business is too complex, and so removed that. Then, I couldn't think of a situation where I would want to exclude genes with a zero in *any* sample, which seems pretty restrictive.

So that brings us down to two options, but with the flexibility to include more options later if it makes sense.

```yaml
  # This sets the `universe` argument of the `enricher` function to choose an option how to limit
  # background genes. Choose one from the following options:
  #   - exclude_zero_in_all: excludes genes with zero raw count in all samples
  #   - no_filter: disable any filtering. This option will use all available genes in each annotation database. You can always use this option and then manually adjust the results objects going into the `enricher` function if you need more customization.
  limit_background_genes: "exclude_zero_in_all"
```

But now that's down to only two options, I'm not sure if it makes sense to have this in the config at all. Furthermore, since the `enrich_test` function already passes its `...` args to `enricher` and `GSEA`, we can already do this in the master branch:

```r
enrich_test(res, TERM2GENE, TERM2NAME, config, direction='up', universe=names_of_genes_to_use)>0])
```

So maybe the solution here is: 1) by default, restrict to nonzero; 2) make it obvious that's what's happening; and 3) possibly provide a helper function to do this.

E.g., 

```r
nonzero_genes <- function(res){
  # something like this
  return names(res)[res$baseMean == 0]
}
enrich_test(res, TERM2GENE, TERM2NAME, config, direction='up', universe_func=nonzero_genes)
```

Reply:
`exclude_zero_in_all` and `no_filter` sound great. 

When it comes to the `enriche_test` function, I think we can keep it as-is as you described. The `enrich_test` is called in the `run_enricher` function. Therefore, one sensible way would be to narrow the `universe` using a wrapper function like your `nonzero_genes` and assign them as input genes before `enrich_test` in the `run_enricher()` function.